### PR TITLE
feat(recurring-job): filesystem trim

### PR DESCRIPTION
**Pull request by c3y1huang**
https://github.com/longhorn/longhorn/issues/5186

### CODE REVIEW

**Comment by c3y1huang** [comment]
Test result: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/3148

**Comment by innobead** [comment]
cc @derekbit https://github.com/longhorn/longhorn/issues/5143

**Comment by innobead** [comment]
@c3y1huang Please fix the conflict.

File: webhook/resources/recurringjob/mutator.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ func (r *recurringJobMutator) Create(request *admission.Request, newObj runtime.
 		"task":         recurringjob.Spec.Task,
 	})
 	switch recurringjob.Spec.Task {
-	case longhorn.RecurringJobTypeSnapshotCleanup:
+	case longhorn.RecurringJobTypeSnapshotCleanup, longhorn.RecurringJobTypeFilesystemTrim:
```
Comment (innobead) [inline webhook/resources/recurringjob/mutator.go:61]: Should do the same in update?

File: app/recurring_job.go [hunk 2]
```diff
@@ -361,18 +361,26 @@ func (job *Job) run() (err error) {
 	}
 
 	// only recurring job types `snapshot` and `backup` need to check if old snapshots can be deleted or not before creating
-	if job.task == longhorn.RecurringJobTypeSnapshot || job.task == longhorn.RecurringJobTypeBackup {
+	switch job.task {
```
Comment (innobead) [inline app/recurring_job.go:364]: It's good to use switch to do multiple values equal comparison.

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, just a few comments needed to resolve.

**Review by shuo-wu — COMMENTED** [review]
Considering the setting/config `Remove Snapshots During Filesystem Trim`, will the recurring job issue a snapshot purge request after a trim request?

**Comment by c3y1huang** [comment]
> Considering the setting/config `Remove Snapshots During Filesystem Trim`, will the recurring job issue a snapshot purge request after a trim request?

Make sense, updated!

File: app/recurring_job.go [hunk 3]
```diff
@@ -506,6 +485,41 @@ func (job *Job) deleteSnapshots(names []string, volume *longhornclient.Volume, v
 	return nil
 }
 
+func (job *Job) purgeSnapshots(volume *longhornclient.Volume, volumeAPI longhornclient.VolumeOperations) error {
+	if _, err := volumeAPI.ActionSnapshotPurge(volume); err != nil {
+		return err
+	}
+	for {
+		done := true
```
Comment (shuo-wu) [inline app/recurring_job.go:493]: Just realized that this is an infinite loop... Should we add a timeout mechanism?

**Review by shuo-wu — COMMENTED** [review]
Others LGTM

File: app/recurring_job.go (hunk 3 above)
Comment (c3y1huang) [reply app/recurring_job.go]: Done!

File: app/recurring_job.go [hunk 4]
```diff
@@ -38,6 +38,7 @@ const (
 	HTTPClientTimout = 1 * time.Minute
 
 	SnapshotPurgeStatusInterval = 5 * time.Second
+	SnapshotPurgeStatusTimeout  = 1 * time.Hour
```
Comment (shuo-wu) [inline app/recurring_job.go:41]: Is 1 hour enough for a large volume purge? I am not sure what the best value is here...

**Review by shuo-wu — APPROVED** [review]
In general LGTM

File: app/recurring_job.go (hunk 4 above)
Comment (c3y1huang) [reply app/recurring_job.go]: Since we're unsure of the appropriate timeout value, I've set it to 24 hours with a comment.

**Review by shuo-wu — APPROVED** [review]

**Review by innobead — APPROVED** [review]
LGTM

**Comment by innobead** [comment]
@c3y1huang make sure other draft PRs ready for review.

**Pull request merged**